File: python/processingformats/locationRequest.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


# package imports
import processingformats.pick

# stdlib imports
import json
import datetime


class LocationRequest:
    """
    A conversion class used to create, parse, and validate location request data
    """

    # JSON Keys
    TYPE_KEY = "Type"  # Required
    ID_KEY = "ID"  # Optional
    SOURCE_KEY = "Source"  # Optional
    EARTHMODEL_KEY = "EarthModel"  # Optional
    SLABRESOLUTION_KEY = "SlabResolution"  # Optional
    SOURCEORIGINTIME_KEY = "SourceOriginTime"  # Required
    SOURCELATITUDE_KEY = "SourceLatitude"  # Required
    SOURCELONGITUDE_KEY = "SourceLongitude"  # Required
    SOURCEDEPTH_KEY = "SourceDepth"  # Required
    INPUTDATA_KEY = "InputData"  # Required
    ISLOCATIONNEW_KEY = "IsLocationNew"  # Optional
    ISLOCATIONHELD_KEY = "IsLocationHeld"  # Optional
    ISDEPTHHELD_KEY = "IsDepthHeld"  # Optional
    ISBAYESIANDEPTH_KEY = "IsBayesianDepth"  # Optional
    BAYESIANDEPTH_KEY = "BayesianDepth"  # Optional
    BAYESIANSPREAD_KEY = "BayesianSpread"  # Optional
    USESVD_KEY = "UseSVD"  # Optional
    REASSESSINITIALPHASEIDS_KEY = "ReassessInitialPhaseIDs"  # Optional
    OUTPUTDATA_KEY = "OutputData"  # Contain output from locator

    # Intialize members
    def __init__(
        self,
        newID=None,
        newSource=None,
        newType=None,
        newEarthModel=None,
        newSlabResolution=None,
        newSourceLatitude=None,
        newSourceLongitude=None,
        newSourceOriginTime=None,
        newSourceDepth=None,
        newInputData=None,
        newIsLocationNew=None,
        newIsLocationHeld=None,
        newIsDepthHeld=None,
        newIsBayesianDepth=None,
        newBayesianDepth=None,
        newBayesianSpread=None,
        newReassessInitialPhaseIDs=None,
        newUseSVD=None,
    ):
        """Initializes the pick object. Constructs empty object if all are none

        newID: a string containing the ID
        newSource: a processingformats.source.Source containing desired source (and supporting info)
        newType: a type identifier for this Location Request
        newEarthModel: an earth model for this Location Request, default ak135
        newSlabResolution: an earth model for this Location Request, default 2spd
        newSourceLatitude: a double containing the source latitude in degrees
        newSourceLongitude: a double containing the source longitude in degrees
        newSourceOriginTime: a Datetime containing the source time
        newSourceDepth: a double containing the source depth in kilometers
        newInputData: a vector of input Pick objects for this Location Result
        newIsLocationNew: a boolean indicating whether the location is new
        newIsLocationHeld: a boolean indicating whether the location is held
        newIsDepthHeld: a boolean indicating whether the depth is held
        newIsBayesianDepth: a boolean indicating whether the depth is bayesian
        newBayesianDepth: a double containing the bayesian depth
        newBayesianSpread: a double containing the bayesian spread
        newUseSVD: a boolean indicating whether use SVD
        newReassessInitialPhaseIDs: a boolean indicating whether reassess inital phase identifications
        """

        self.outputData = None

        # Required Keys
        if newType is not None:
            self.type = newType

        if newSourceLatitude is not None:
            self.sourceLatitude = newSourceLatitude

        if newSourceLongitude is not None:
            self.sourceLongitude = newSourceLongitude

        if newSourceDepth is not None:
            self.sourceDepth = newSourceDepth

        if newSourceOriginTime is not None:
            self.sourceOriginTime = newSourceOriginTime

        if newInputData is not None:
            self.inputData = newInputData

        # Optional Keys
        if newID is not None:
            self.id = newID

        if newSource is not None:
            self.source = newSource
        else:
            self.source = processingformats.source.Source()

        if newEarthModel is not None:
            self.earthModel = newEarthModel
        else:
            self.earthModel = "ak135"

        if newSlabResolution is not None:
            self.slabResolution = newSlabResolution
        else:
            self.slabResolution = "2spd"

        if newIsLocationNew is not None:
            self.isLocationNew = newIsLocationNew

        if newIsLocationHeld is not None:
            self.isLocationHeld = newIsLocationHeld

        if newIsDepthHeld is not None:
            self.isDepthHeld = newIsDepthHeld

        if newIsBayesianDepth is not None:
            self.isBayesianDepth = newIsBayesianDepth

        if newBayesianDepth is not None:
            self.bayesianDepth = newBayesianDepth

        if newBayesianSpread is not None:
            self.bayesianSpread = newBayesianSpread

        if newUseSVD is not None:
            self.useSVD = newUseSVD

        if newReassessInitialPhaseIDs is not None:
            self.reassessInitialPhaseIDs = newReassessInitialPhaseIDs

    def fromJSONString(self, JSONString):
        """Populates object from a JSON formatted string

        JSONString: a required string containing the JSON formatted text
        """
        JSONObject = json.loads(JSONString)
        self.fromDict(JSONObject)

    def fromDict(self, aDict):
        """Populates object from a dictionary

        aDict: a required dictionary
        """

        # Required Keys
        try:
            self.type = aDict[self.TYPE_KEY]
            self.sourceLatitude = aDict[self.SOURCELATITUDE_KEY]
            self.sourceLongitude = aDict[self.SOURCELONGITUDE_KEY]
            self.sourceDepth = aDict[self.SOURCEDEPTH_KEY]
            timeString = aDict[self.SOURCEORIGINTIME_KEY][:-1] + "000Z"
            self.sourceOriginTime = datetime.datetime.strptime(
                timeString, "%Y-%m-%dT%H:%M:%S.%fZ"
            )

            if self.INPUTDATA_KEY in aDict:
                aDataList = aDict[self.INPUTDATA_KEY]
                if aDataList:
                    self.inputData = []

                    for aData in aDataList:
                        newPick = processingformats.pick.Pick()
                        newPick.fromDict(aData)
                        self.inputData.append(newPick)

        except (ValueError, KeyError, TypeError) as e:
            print("Dictionary format error, missing required keys: %s" % e)

        # Optional Keys
        if self.ID_KEY in aDict:
            self.id = aDict[self.ID_KEY]

        if self.SOURCE_KEY in aDict:
            self.source.fromDict(aDict[self.SOURCE_KEY])

        if self.EARTHMODEL_KEY in aDict:
            self.earthModel = aDict[self.EARTHMODEL_KEY]
        else:
            self.earthModel = "ak135"

        if self.SLABRESOLUTION_KEY in aDict:
            self.slabResolution = aDict[self.SLABRESOLUTION_KEY]
        else:
            self.slabResolution = "2spd"

        if self.ISLOCATIONNEW_KEY in aDict:
            self.isLocationNew = aDict[self.ISLOCATIONNEW_KEY]

        if self.ISLOCATIONHELD_KEY in aDict:
            self.isLocationHeld = aDict[self.ISLOCATIONHELD_KEY]

        if self.ISDEPTHHELD_KEY in aDict:
            self.isDepthHeld = aDict[self.ISDEPTHHELD_KEY]

        if self.ISBAYESIANDEPTH_KEY in aDict:
            self.isBayesianDepth = aDict[self.ISBAYESIANDEPTH_KEY]

        if self.BAYESIANDEPTH_KEY in aDict:
            self.bayesianDepth = aDict[self.BAYESIANDEPTH_KEY]

        if self.BAYESIANSPREAD_KEY in aDict:
            self.bayesianSpread = aDict[self.BAYESIANSPREAD_KEY]

        if self.USESVD_KEY in aDict:
            self.useSVD = aDict[self.USESVD_KEY]

        if self.REASSESSINITIALPHASEIDS_KEY in aDict:
            self.reassessInitialPhaseIDs = aDict[self.REASSESSINITIALPHASEIDS_KEY]

    def toJSONString(self):
        """Converts object to a JSON formatted string

        Returns: JSON formatted message as a string
        """
        JSONObject = self.toDict()

        return json.dumps(JSONObject, ensure_ascii=False)

    def toDict(self):
        """Converts object to a dictionary

        Returns: the dictionary
        """

        aDict = {}

        # Required Keys
        try:
            aDict[self.TYPE_KEY] = self.type
            aDict[self.SOURCELATITUDE_KEY] = self.sourceLatitude
            aDict[self.SOURCELONGITUDE_KEY] = self.sourceLongitude
            aDict[self.SOURCEDEPTH_KEY] = self.sourceDepth
            timestring = self.sourceOriginTime.isoformat(timespec="milliseconds") + "Z"
            aDict[self.SOURCEORIGINTIME_KEY] = timestring

            aDataList = []
            if self.inputData and len(self.inputData) > 0:
                for aData in self.inputData:
                    aDataList.append(aData.toDict())

            aDict[self.INPUTDATA_KEY] = aDataList
        except (NameError, AttributeError) as e:
            print("Missing required data error: %s" % e)

        # Optional Keys
        if hasattr(self, "id"):
            aDict[self.ID_KEY] = self.id

        if hasattr(self, "source"):
            if not self.source.isEmpty():
                aDict[self.SOURCE_KEY] = self.source.toDict()

        if hasattr(self, "earthModel"):
            aDict[self.EARTHMODEL_KEY] = self.earthModel
        else:
            aDict[self.EARTHMODEL_KEY] = "ak135"

        if hasattr(self, "slabResolution"):
            aDict[self.SLABRESOLUTION_KEY] = self.slabResolution
        else:
            aDict[self.SLABRESOLUTION_KEY] = "2spd"

        if hasattr(self, "isLocationNew"):
            aDict[self.ISLOCATIONNEW_KEY] = self.isLocationNew

        if hasattr(self, "isLocationHeld"):
            aDict[self.ISLOCATIONHELD_KEY] = self.isLocationHeld

        if hasattr(self, "isDepthHeld"):
            aDict[self.ISDEPTHHELD_KEY] = self.isDepthHeld

        if hasattr(self, "isBayesianDepth"):
            aDict[self.ISBAYESIANDEPTH_KEY] = self.isBayesianDepth

        if hasattr(self, "bayesianDepth"):
            aDict[self.BAYESIANDEPTH_KEY] = self.bayesianDepth

        if hasattr(self, "bayesianSpread"):
            aDict[self.BAYESIANSPREAD_KEY] = self.bayesianSpread

        if hasattr(self, "useSVD"):
            aDict[self.USESVD_KEY] = self.useSVD

        if hasattr(self, "reassessInitialPhaseIDs"):
            aDict[self.REASSESSINITIALPHASEIDS_KEY] = self.reassessInitialPhaseIDs

        return aDict

    def isValid(self):
        """Checks to see if object is valid

        Returns: true if object is valid, false otherwise
        """
        errorList = self.getErrors()

        return not errorList

    def getErrors(self):
        """Gets a list of object validation errors

        Returns: a list of string containing the validation error messages
        """
        errorList = []

        # required keys

        # sourceLatitude
        try:
            if self.sourceLatitude < -90 or self.sourceLatitude > 90:
                errorList.append(
                    "Source Latitude in LocationRequest Class not in the range of -90 to 90."
                )
        except (NameError, AttributeError):
            errorList.append("No Source Latitude in LocationRequest Class.")

        # sourceLongitude
        try:
            if self.sourceLongitude < -180 or self.sourceLongitude > 180:
                errorList.append(
                    "Source Longitude in LocationRequest Class not in range of -180 to 180."
                )
        except (NameError, AttributeError):
            errorList.append("No Source Longitude in LocationRequest Class")

        # sourceDepth
        try:
            if self.sourceDepth < -100 or self.sourceDepth > 1500:
                errorList.append(
                    "Source Depth in LocationRequest Class not in the range of -100 to 1500"
                )
        except (NameError, AttributeError):
            errorList.append("No Source Depth in LocationRequest Class.")

        # sourceOriginTime
        try:
            self.sourceOriginTime
        except (NameError, AttributeError):
            errorList.append("No Source Origin Time in LocationRequest Class.")

        # inputData
        if hasattr(self, "inputData"):
            if self.inputData and len(self.inputData) > 0:
                for anInput in self.inputData:
                    if not anInput.isValid():
                        errorList.append("Invalid Input in LocationRequest Class.")

        if hasattr(self, "source"):
            if not self.source.isEmpty():
                if not self.source.isValid():
                    errorList.append("Invalid Source in LocationRequest Class.")

        return errorList

File: cpp/src/LocationRequest.cpp
#include <LocationRequest.h>

#include <string>
#include <limits>
#include <vector>

// JSON Keys
#define TYPE_KEY "Type"
#define SOURCE_KEY "Source"
#define ID_KEY "ID"
#define EARTHMODEL_KEY "EarthModel"
#define SLABRESOLUTION_KEY "SlabResolution"
#define SOURCEORIGINTIME_KEY "SourceOriginTime"
#define SOURCELATITUDE_KEY "SourceLatitude"
#define SOURCELONGITUDE_KEY "SourceLongitude"
#define SOURCEDEPTH_KEY "SourceDepth"
#define INPUTDATA_KEY "InputData"
#define ISLOCATIONNEW_KEY "IsLocationNew"
#define ISLOCATIONHELD_KEY "IsLocationHeld"
#define ISDEPTHHELD_KEY "IsDepthHeld"
#define ISBAYESIANDEPTH_KEY "IsBayesianDepth"
#define BAYESIANDEPTH_KEY "BayesianDepth"
#define BAYESIANSPREAD_KEY "BayesianSpread"
#define USESVD_KEY "UseSVD"
#define OUTPUTDATA_KEY "OutputData"

namespace processingformats {
LocationRequest::LocationRequest() {
	id = "";
    source = processingformats::Source();
    type = "";
    earthModel = "ak135";
	slabResolution = "20spd";
    sourceLatitude = std::numeric_limits<double>::quiet_NaN();
    sourceLongitude = std::numeric_limits<double>::quiet_NaN();
    sourceOriginTime = std::numeric_limits<double>::quiet_NaN();
    sourceDepth = std::numeric_limits<double>::quiet_NaN();
    inputData.clear();
    isLocationNew = false;
    isLocationHeld = false;
    isDepthHeld = false;
    isBayesianDepth = false;
    bayesianDepth = std::numeric_limits<double>::quiet_NaN();
    bayesianSpread = std::numeric_limits<double>::quiet_NaN();
    useSVD = false;
    outputData = processingformats::LocationResult();
}

LocationRequest::LocationRequest(
      std::string newID,
      std::string newAgencyID,
      std::string newAuthor,
      std::string newType,
      std::string newLocType,
      std::string newEarthModel,
	  std::string newSlabResolution,
      double newSourceLatitude,
      double newSourceLongitude,
      double newSourceOriginTime,
      double newSourceDepth,
      std::vector<processingformats::Pick> newInputData,
      bool newIsLocationNew,
      bool newIsLocationHeld,
      bool newIsDepthHeld,
      bool newIsBayesianDepth,
      double newBayesianDepth,
      double newBayesianSpread,
      bool newUseSVD) {
	id = newID;
	LocationRequest::source = processingformats::Source(newAgencyID, newAuthor,
		newType);
	type = newLocType;
	sourceLatitude = newSourceLatitude;
	sourceLongitude = newSourceLongitude;
	sourceOriginTime = newSourceOriginTime;
	sourceDepth = newSourceDepth;
	inputData.clear();
	for (int i = 0; i < static_cast<int>(newInputData.size()); i++) {
		inputData.push_back(newInputData[i]);
	}
	if (newEarthModel != "") {
		earthModel = newEarthModel;
	}
	if (newSlabResolution != "") {
		slabResolution= newSlabResolution;
	}
	isLocationNew = newIsLocationNew;
	isLocationHeld = newIsLocationHeld;
	isDepthHeld = newIsDepthHeld;
	isBayesianDepth = newIsBayesianDepth;
	bayesianDepth = newBayesianDepth;
	bayesianSpread = newBayesianSpread;
	useSVD = newUseSVD;
	outputData = processingformats::LocationResult();
}

LocationRequest::LocationRequest(
      std::string newID,
      processingformats::Source newSource,
      std::string newLocType,
      std::string newEarthModel,
	  std::string newSlabResolution,
      double newSourceLatitude,
      double newSourceLongitude,
      double newSourceOriginTime,
      double newSourceDepth,
      std::vector<processingformats::Pick> newInputData,
      bool newIsLocationNew,
      bool newIsLocationHeld,
      bool newIsDepthHeld,
      bool newIsBayesianDepth,
      double newBayesianDepth,
      double newBayesianSpread,
      bool newUseSVD) {
	id = newID;
	LocationRequest::source = newSource;
	type = newLocType;
	sourceLatitude = newSourceLatitude;
	sourceLongitude = newSourceLongitude;
	sourceOriginTime = newSourceOriginTime;
	sourceDepth = newSourceDepth;
	inputData.clear();
	for (int i = 0; i < static_cast<int>(newInputData.size()); i++) {
		inputData.push_back(newInputData[i]);
	}
	if (newEarthModel != "") {
		earthModel = newEarthModel;
	}
	if (newSlabResolution != "") {
		slabResolution= newSlabResolution;
	}  
	isLocationNew = newIsLocationNew;
	isLocationHeld = newIsLocationHeld;
	isDepthHeld = newIsDepthHeld;
	isBayesianDepth = newIsBayesianDepth;
	bayesianDepth = newBayesianDepth;
	bayesianSpread = newBayesianSpread;
	useSVD = newUseSVD;
	outputData = processingformats::LocationResult();
}

LocationRequest::LocationRequest(rapidjson::Value &json) {
	// required values
	// type
	if ((json.HasMember(TYPE_KEY) == true)
			&& (json[TYPE_KEY].IsString() == true)) {
		type = std::string(json[TYPE_KEY].GetString(),
								json[TYPE_KEY].GetStringLength());
	} else {
		type = "";
	}

    // sourceLatitude
	if ((json.HasMember(SOURCELATITUDE_KEY) == true)
			&& (json[SOURCELATITUDE_KEY].IsNumber() == true)
			&& (json[SOURCELATITUDE_KEY].IsDouble() == true)) {
		sourceLatitude = json[SOURCELATITUDE_KEY].GetDouble();
	} else {
		sourceLatitude = std::numeric_limits<double>::quiet_NaN();
	}

	// sourceLongitude
	if ((json.HasMember(SOURCELONGITUDE_KEY) == true)
			&& (json[SOURCELONGITUDE_KEY].IsNumber() == true)
			&& (json[SOURCELONGITUDE_KEY].IsDouble() == true)) {
		sourceLongitude = json[SOURCELONGITUDE_KEY].GetDouble();
	} else {
		sourceLongitude = std::numeric_limits<double>::quiet_NaN();
	}

	// sourceDepth
	if ((json.HasMember(SOURCEDEPTH_KEY) == true)
			&& (json[SOURCEDEPTH_KEY].IsNumber() == true)
			&& (json[SOURCEDEPTH_KEY].IsDouble() == true)) {
		sourceDepth = json[SOURCEDEPTH_KEY].GetDouble();
	} else {
		sourceDepth = std::numeric_limits<double>::quiet_NaN();
	}

    // sourceOriginTime
	if ((json.HasMember(SOURCEORIGINTIME_KEY) == true)
			&& (json[SOURCEORIGINTIME_KEY].IsString() == true)) {
		sourceOriginTime = processingformats::ConvertISO8601ToEpochTime(
				std::string(json[SOURCEORIGINTIME_KEY].GetString(),
							json[SOURCEORIGINTIME_KEY].GetStringLength()));
	} else {
		sourceOriginTime = std::numeric_limits<double>::quiet_NaN();
	}

    // input data
	inputData.clear();
	if ((json.HasMember(INPUTDATA_KEY) == true)
			&& (json[INPUTDATA_KEY].IsArray() == true)) {
		rapidjson::Value dataarray;
		dataarray = json[INPUTDATA_KEY].GetArray();

		for (rapidjson::SizeType i = 0; i < dataarray.Size(); i++) {
			rapidjson::Value & datavalue = dataarray[i];

			processingformats::Pick newdata(datavalue);

			// add to vector
			inputData.push_back(newdata);
		}
	}

	// optional values
	// id
	// check type of id object
	if ((json.HasMember(ID_KEY) == true) && (json[ID_KEY].IsString() == true)) {
		// the Id *should* be a string
		id = std::string(json[ID_KEY].GetString(),
							json[ID_KEY].GetStringLength());
	} else if ((json.HasMember(ID_KEY) == true) 
		&& (json[ID_KEY].IsNumber() == true) 
		&& (json[ID_KEY].IsInt() == true)) {
		// but the ID *could* be an int, but we 
        // want it to be a string
		id = std::to_string(json[SOURCELONGITUDE_KEY].GetInt());
	} else {
		// any other type isn't a usable id
		id = "";
	}

	// source
	if ((json.HasMember(SOURCE_KEY) == true)
			&& (json[SOURCE_KEY].IsObject() == true)) {
		rapidjson::Value & sourcevalue = json[SOURCE_KEY];
		source = processingformats::Source(sourcevalue);
	} else {
		source = processingformats::Source();
	}

	// earthModel
	if ((json.HasMember(EARTHMODEL_KEY) == true)
			&& (json[EARTHMODEL_KEY].IsString() == true)) {
		earthModel = std::string(json[EARTHMODEL_KEY].GetString(),
								json[EARTHMODEL_KEY].GetStringLength());
	} else {
		earthModel = "ak135";
	}

	// slabResolution
	if ((json.HasMember(SLABRESOLUTION_KEY) == true)
			&& (json[SLABRESOLUTION_KEY].IsString() == true)) {
		slabResolution = std::string(json[SLABRESOLUTION_KEY].GetString(),
								json[SLABRESOLUTION_KEY].GetStringLength());
	} else {
		slabResolution = "20spd";
	}

  	// isLocationNew
	if ((json.HasMember(ISLOCATIONNEW_KEY) == true)
			&& (json[ISLOCATIONNEW_KEY].IsBool() == true)) {
		isLocationNew = json[ISLOCATIONNEW_KEY].GetBool();
	} else {
		isLocationNew = false;
	}

	// isLocationHeld
	if ((json.HasMember(ISLOCATIONHELD_KEY) == true)
			&& (json[ISLOCATIONHELD_KEY].IsBool() == true)) {
		isLocationHeld = json[ISLOCATIONHELD_KEY].GetBool();
	} else {
		isLocationHeld = false;
	}

	// isDepthHeld
	if ((json.HasMember(ISDEPTHHELD_KEY) == true)
			&& (json[ISDEPTHHELD_KEY].IsBool() == true)) {
		isDepthHeld = json[ISDEPTHHELD_KEY].GetBool();
	} else {
		isDepthHeld = false;
	}

	// isBayesianDepth
	if ((json.HasMember(ISBAYESIANDEPTH_KEY) == true)
			&& (json[ISBAYESIANDEPTH_KEY].IsBool() == true)) {
		isBayesianDepth = json[ISBAYESIANDEPTH_KEY].GetBool();
	} else {
		isBayesianDepth = false;
	}

	// bayesianDepth
	if ((json.HasMember(BAYESIANDEPTH_KEY) == true)
			&& (json[BAYESIANDEPTH_KEY].IsNumber() == true)
			&& (json[BAYESIANDEPTH_KEY].IsDouble() == true)) {
		bayesianDepth = json[BAYESIANDEPTH_KEY].GetDouble();
	} else {
		bayesianDepth = std::numeric_limits<double>::quiet_NaN();
	}

	// bayesianSpread
	if ((json.HasMember(BAYESIANSPREAD_KEY) == true)
			&& (json[BAYESIANSPREAD_KEY].IsNumber() == true)
			&& (json[BAYESIANSPREAD_KEY].IsDouble() == true)) {
		bayesianSpread = json[BAYESIANSPREAD_KEY].GetDouble();
	} else {
		bayesianSpread = std::numeric_limits<double>::quiet_NaN();
	}

	// useSVD
	if ((json.HasMember(USESVD_KEY) == true)
			&& (json[USESVD_KEY].IsBool() == true)) {
		useSVD = json[USESVD_KEY].GetBool();
	} else {
		useSVD = false;
	}

	// outputData
	if ((json.HasMember(OUTPUTDATA_KEY) == true)
			&& (json[OUTPUTDATA_KEY].IsObject() == true)) {
		rapidjson::Value & sourcevalue = json[OUTPUTDATA_KEY];
		outputData = processingformats::LocationResult(sourcevalue);
	} else {
		outputData = processingformats::LocationResult();
	}
}

LocationRequest::LocationRequest(const LocationRequest & newLocationRequest) {
	  id = newLocationRequest.id;
    source = newLocationRequest.source;
    type = newLocationRequest.type;
    sourceLatitude = newLocationRequest.sourceLatitude;
    sourceLongitude = newLocationRequest.sourceLongitude;
    sourceOriginTime = newLocationRequest.sourceOriginTime;
    sourceDepth = newLocationRequest.sourceDepth;

    // copy data
    inputData.clear();
    for (int i = 0; i < static_cast<int>(newLocationRequest.inputData.size());
        i++) {
		  inputData.push_back(newLocationRequest.inputData[i]);
	  }

    earthModel = newLocationRequest.earthModel;
	slabResolution = newLocationRequest.slabResolution;
    isLocationNew = newLocationRequest.isLocationNew;
    isLocationHeld = newLocationRequest.isLocationHeld;
    isDepthHeld = newLocationRequest.isDepthHeld;
    isBayesianDepth = newLocationRequest.isBayesianDepth;
    bayesianDepth = newLocationRequest.bayesianDepth;
    bayesianSpread = newLocationRequest.bayesianSpread;
    useSVD = newLocationRequest.useSVD;
    outputData = newLocationRequest.outputData;
}

LocationRequest::~LocationRequest() {
	inputData.clear();
}

rapidjson::Value & LocationRequest::toJSON(
		rapidjson::Value &json,
		rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> &allocator) {
	json.SetObject();

	// required values
	// type
	if (type != "") {
		rapidjson::Value typeValue;
		typeValue.SetString(rapidjson::StringRef(type.c_str()),
								allocator);
		json.AddMember(TYPE_KEY, typeValue, allocator);
	}

	// sourceLatitude
	if (std::isnan(sourceLatitude) != true) {
		json.AddMember(SOURCELATITUDE_KEY, sourceLatitude, allocator);
	}

	// sourceLongitude
	if (std::isnan(sourceLongitude) != true) {
		json.AddMember(SOURCELONGITUDE_KEY, sourceLongitude, allocator);
  	}

	// sourceDepth
	if (std::isnan(sourceDepth) != true) {
		json.AddMember(SOURCEDEPTH_KEY, sourceDepth, allocator);
  	}

	// sourceOriginTime
	if (std::isnan(sourceOriginTime) != true) {
		std::string timestring = processingformats::ConvertEpochTimeToISO8601(
				sourceOriginTime);
		rapidjson::Value timevalue;
		timevalue.SetString(rapidjson::StringRef(timestring.c_str()),
							allocator);
		json.AddMember(SOURCEORIGINTIME_KEY, timevalue, allocator);
	}

  	// input data
	rapidjson::Value dataarray(rapidjson::kArrayType);
	if (inputData.size() > 0) {
		for (int i = 0; i < static_cast<int>(inputData.size()); i++) {
			rapidjson::Value datavalue(rapidjson::kObjectType);
			inputData[i].toJSON(datavalue, allocator);
			dataarray.PushBack(datavalue, allocator);
		}
	}

	if (dataarray.Size() > 0) {
		// data
		json.AddMember(INPUTDATA_KEY, dataarray, allocator);
	}

  	// Optional values
	// id
	if (id != "") {
		rapidjson::Value idValue;
		idValue.SetString(rapidjson::StringRef(id.c_str()), allocator);
		json.AddMember(ID_KEY, idValue, allocator);
	}

	// source
	if (source.isEmpty() == false) {
		rapidjson::Value sourceValue(rapidjson::kObjectType);
		source.toJSON(sourceValue, allocator);
		json.AddMember(SOURCE_KEY, sourceValue, allocator);
	}

	// earthModel
	rapidjson::Value earthModelValue;
	earthModelValue.SetString(rapidjson::StringRef(earthModel.c_str()),
							allocator);
	json.AddMember(EARTHMODEL_KEY, earthModelValue, allocator);

	// slabResolution
	rapidjson::Value slabResolutionValue;
	slabResolutionValue.SetString(rapidjson::StringRef(slabResolution.c_str()),
							allocator);
	json.AddMember(SLABRESOLUTION_KEY, slabResolutionValue, allocator);

	// isLocationNew
	json.AddMember(ISLOCATIONNEW_KEY, isLocationNew, allocator);

	// isLocationHeld
	json.AddMember(ISLOCATIONHELD_KEY, isLocationHeld, allocator);

	// isDepthHeld
	json.AddMember(ISDEPTHHELD_KEY, isDepthHeld, allocator);

	// isBayesianDepth
	json.AddMember(ISBAYESIANDEPTH_KEY, isBayesianDepth, allocator);

	// bayesianDepth
	if (std::isnan(bayesianDepth) != true) {
		json.AddMember(BAYESIANDEPTH_KEY, bayesianDepth, allocator);
  	}

	// bayesianSpread
	if (std::isnan(bayesianSpread) != true) {
		json.AddMember(BAYESIANSPREAD_KEY, bayesianSpread, allocator);
  	}

	// useSVD
	json.AddMember(USESVD_KEY, useSVD, allocator);

	// outputData
	if (outputData.isEmpty() == false) {
		rapidjson::Value outputDataValue(rapidjson::kObjectType);
		outputData.toJSON(outputDataValue, allocator);
		json.AddMember(OUTPUTDATA_KEY, outputDataValue, allocator);
	}

	return (json);
}

std::vector<std::string> LocationRequest::getErrors() {
	std::vector<std::string> errorlist;

	// check required data
	// sourceLatitude
	if (std::isnan(sourceLatitude) == true) {
		// sourceLatitude not found
		errorlist.push_back("No sourceLatitude in LocationRequest class.");
	} else if ((sourceLatitude < -90.0) || (sourceLatitude > 90.0)) {
		errorlist.push_back("Invalid sourceLatitude in LocationRequest class.");
	}

	// sourceLongitude
	if (std::isnan(sourceLongitude) == true) {
		// sourceLongitude not found
		errorlist.push_back("No sourceLongitude in LocationRequest class.");
	} else if ((sourceLongitude < -180.0) || (sourceLongitude > 180.0)) {
		errorlist.push_back("Invalid sourceLongitude in LocationRequest class.");
	}

	// sourceOriginTime
	if (std::isnan(sourceOriginTime) == true) {
		errorlist.push_back("sourceOriginTime is missing in LocationRequest class.");
	} else {
		try {
			if (processingformats::IsStringISO8601(
					processingformats::ConvertEpochTimeToISO8601(sourceOriginTime))
					== false) {
				errorlist.push_back(
						"sourceOriginTime did not validate in LocationRequest class.");
			}
		} catch (const std::exception & e) {
			errorlist.push_back(std::string(e.what()));
		}
	}

	// sourceDepth
	if (std::isnan(sourceDepth) == true) {
		// sourceDepth not found
		errorlist.push_back("No sourceDepth in LocationRequest class.");
	} else if ((sourceDepth < -100) || (sourceDepth > 1500)) {
		errorlist.push_back("Invalid sourceDepth in LocationRequest class.");
	}

  	// inputData
	if (inputData.size() > 0) {
		for (int i = 0; i < static_cast<int>(inputData.size()); i++) {
			if (inputData[i].isValid() != true) {
				std::vector<std::string> dataErrors = inputData[i].getErrors();

				std::string errorString =
						"Invalid input data in LocationRequest class:";

				for (int j = 0; j < dataErrors.size(); j++) {
					errorString += " " + dataErrors[j];
				}

				// bad pick
				errorlist.push_back(errorString);
			}
		}
	}

  	// optional values
	// source
	if (source.isEmpty() == false) {
		if (source.isValid() != true) {
			std::vector<std::string> sourceErrors = source.getErrors();

			std::string errorString =
				"Source object did not validate in LocationRequest class:";

			for (int i = 0; i < sourceErrors.size(); i++) {
				errorString += " " + sourceErrors[i];
			}

			// bad source
			errorlist.push_back(errorString);
		}
	}

	// outputData
	if (outputData.isEmpty() == false) {
		if (outputData.isValid() != true) {
			std::vector<std::string> outputDataErrors = outputData.getErrors();

			std::string errorString =
					"outputData object did not validate in LocationRequest class:";

			for (int i = 0; i < outputDataErrors.size(); i++) {
				errorString += " " + outputDataErrors[i];
			}

			// bad outputData
			errorlist.push_back(errorString);
		}
	}

	// return the list of errors
	return (errorlist);
}
}  // namespace processingformats
